engine/src/main/java/org/pentaho/di/trans/steps/calculator/Calculator.java

```diff
@@ -0,0 +1,664 @@
+            break;
+          case CalculatorMetaFunction.CALC_DATE_DIFF_HR: // DateA - DateB (h)
+            calcData[index] = ValueDataUtil.DateDiff( metaA, dataA, metaB, dataB, "h" );
+            resultType = CalculatorMetaFunction.calcDefaultResultType[calcType];
+            break;
+          case CalculatorMetaFunction.CALC_HOUR_OF_DAY:
+            calcData[index] = ValueDataUtil.hourOfDay( metaA, dataA );
+            resultType = CalculatorMetaFunction.calcDefaultResultType[calcType];
+            break;
+          case CalculatorMetaFunction.CALC_MINUTE_OF_HOUR:
+            calcData[index] = ValueDataUtil.minuteOfHour( metaA, dataA );
+            resultType = CalculatorMetaFunction.calcDefaultResultType[calcType];
+            break;
+          case CalculatorMetaFunction.CALC_SECOND_OF_MINUTE:
+            calcData[index] = ValueDataUtil.secondOfMinute( metaA, dataA );
+            resultType = CalculatorMetaFunction.calcDefaultResultType[calcType];
+            break;
+          case CalculatorMetaFunction.CALC_ADD_SECONDS: // Add B seconds to date field A
+            calcData[index] = ValueDataUtil.addSeconds( metaA, dataA, metaB, dataB );
+            resultType = CalculatorMetaFunction.calcDefaultResultType[calcType];
+            break;
+          case CalculatorMetaFunction.CALC_REMAINDER:
+            if ( targetMeta.getType() != metaA.getType() || targetMeta.getType() != metaB.getType() ) {
+              dataA = targetMeta.convertData( metaA, dataA );
+              metaA = targetMeta.clone();
+              dataB = targetMeta.convertData( metaB, dataB );
+              metaB = targetMeta.clone();
+            }
+            calcData[index] = ValueDataUtil.remainder( metaA, dataA, metaB, dataB );
+            resultType = targetMeta.getType();
+            break;
+          default:
+            throw new KettleValueException( BaseMessages.getString( PKG, "Calculator.Log.UnknownCalculationType" )
+              + fn.getCalcType() );
+        }
+
+        // If we don't have a target data type, throw an error.
+        // Otherwise the result is non-deterministic.
+        //
+        if ( targetMeta.getType() == ValueMetaInterface.TYPE_NONE ) {
+          throw new KettleValueException( BaseMessages.getString( PKG, "Calculator.Log.NoType" )
+            + ( i + 1 ) + " : " + fn.getFieldName() + " = " + fn.getCalcTypeDesc() + " / "
+            + fn.getCalcTypeLongDesc() );
+        }
+
+        // Convert the data to the correct target data type.
+        //
+        if ( calcData[index] != null ) {
+          if ( targetMeta.getType() != resultType ) {
+            ValueMetaInterface resultMeta;
+            try {
+              // clone() is not necessary as one data instance belongs to one step instance and no race condition occurs
+              resultMeta = data.getValueMetaFor( resultType, "result" );
+            } catch ( Exception exception ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)